### PR TITLE
NAS-119534 / 23.10 / Update Linux kernel to v6.1.24 for Cobia

### PR DESCRIPTION
**Pull request by umsaleem**
All of our local patches for TrueNAS SCALE are ported on top of v6.1.24, except for https://github.com/truenas/linux/pull/17, which can be added later.

### CODE REVIEW

File: drivers/nvdimm/pmem.c [hunk 1]
```diff
@@ -31,6 +33,24 @@
 #include "pfn.h"
 #include "nd.h"
```
Comment (umsaleem) [inline drivers/nvdimm/pmem.c:35]: There were many conflicts in this file. I did not patch anything here but mostly updates were made here from upstream and patch was not applying cleanly.

File: drivers/scsi/ses.c [hunk 2]
```diff
@@ -444,10 +448,20 @@ static int ses_process_descriptor(struct enclosure_component *ecomp,
 	struct ses_component *scomp = ecomp->scratch;
```
Comment (umsaleem) [inline drivers/scsi/ses.c:448]: Return type of this function was changed from void to int.

File: fs/attr.c [hunk 3]
```diff
@@ -182,21 +182,69 @@ int setattr_prepare(struct user_namespace *mnt_userns, struct dentry *dentry,
...
 
 	/* Make sure a caller can chown. */
+#if CONFIG_TRUENAS
+	/*
+	 * Check for ACE4_WRITE_OWNER. RFC 5661 Section 6.2.1.3.1
+	 * On UNIX systems, this is the ability to execute chown() and
+	 * chgrp().
+	 */
+	if ((ia_valid & ATTR_UID) &&
+	    !chown_ok(mnt_userns, inode, attr->ia_vfsuid)) {
```
Comment (umsaleem) [inline fs/attr.c:192]: chown_ok interface was changed, takes `vfsuid_t` in place of `kuid_t`

File: fs/attr.c [hunk 4]
```diff
@@ -182,21 +182,69 @@ int setattr_prepare(struct user_namespace *mnt_userns, struct dentry *dentry,
...
+#else
 	if ((ia_valid & ATTR_UID) &&
 	    !chown_ok(mnt_userns, inode, attr->ia_vfsuid))
 		return -EPERM;
+#endif
 
 	/* Make sure caller can chgrp. */
+#if CONFIG_TRUENAS
+	if ((ia_valid & ATTR_GID) &&
+	    !chgrp_ok(mnt_userns, inode, attr->ia_vfsgid)) {
```
Comment (umsaleem) [inline fs/attr.c:209]: chgrp_ok interface was changed, takes `vfsgid_t` in place of `kgid_t`

File: drivers/ata/ahciem.c [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+static int ahciem_queuecommand(struct Scsi_Host *shost, struct scsi_cmnd *cmd)
+{
+	struct ahciem_enclosure *enc = (struct ahciem_enclosure *)&shost->hostdata[0];
+	struct ahciem_args args = { .cmd = cmd, .enc = enc };
+	const u8 *cdb = cmd->cmnd;
+
+	if (unlikely(!cmd->cmd_len)) {
+		printk("ahciem: bad CDB len=0 in %s\n", DRV_NAME);
+		cmd->result = DID_ERROR << 16;
+		scsi_done(cmd);
```
Comment (umsaleem) [inline drivers/ata/ahciem.c:433]: This line was modified. Previously, `scsi_cmnd` struct contained a function pointer for scsi_done callback. But now it is decalared as a function in scsi.h.

File: drivers/ata/ahciem.c [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+			break;
+		}
+		break;
+
+	default:
+		ahciem_scsi_set_sense(cmd, ILLEGAL_REQUEST, 0x20, 0x0);
+		break;
+	}
+
+	scsi_done(cmd);
```
Comment (umsaleem) [inline drivers/ata/ahciem.c:515]: This line was modified. Previously, `scsi_cmnd` struct contained a function pointer for scsi_done callback. But now it is decalared as a function in scsi.h.

**Review by ixhamza — APPROVED** [review]

**Review by amotin — APPROVED** [review]

**Pull request merged**